### ISSUE TITLE
Move 2FA enforcement behind a Waffle Flag instead of a Switch

### ISSUE THREAD

**Issue by diox** [mozilla/addons-server#14972]
2FA enforcement for developers is tracked in https://github.com/mozilla/addons/issues/732. Although the upstream work in FxA that needed to happen to unblock it has been done, there are bugs in the implementation we need to investigate before enabling it.

We moved dev to use a more reliable FxA environment in https://github.com/mozilla/addons-server/issues/14949 so now, we need to make it easier to test on dev. The idea is to a) change our code to make it follow the latest FxA parameters that need to be sent b) move that behind a waffle flag and not a waffle switch, so have more granularity around who that feature gets enabled for.

**Comment by diox** [comment]
**QA:** this is a bit special because I do expect some things to be broken with the flag enabled. For this reason, you need to test both with and without the `2fa-enforcement-for-developers-and-special-users` flag being enabled *for your user(s)*. To do this, go to the django admin and create/edit a waffle *Flag* (not Switch or Sample) named `2fa-enforcement-for-developers-and-special-users`, and add your user id to `Users` at the bottom. Do not activate use other properties of the flag for now, it needs to be per-user for the time being to prevent breaking login for everyone.

With the waffle flag not enabled for your user or completely absent, everything should work transparently as before. You should be able to log in without any problems, whether you're a basic user, a developer, someone with more permissions. If that all works, you can consider the issue verified.

**Comment by diox** [comment]
With the waffle flag enabled, as mentioned above I do expect issues but you can try the following STRs to help us debug things out:

- Create an user, either submit an add-on or add that user to a group through the admin (any group, you can even create a special group just for them in the admin)
- Active the waffle flag '2fa-enforcement-for-developers-and-special-users' for that user through the admin
- Log out
- Try to log back in

Expected results:
- When logging back in, after submitting your password, FxA should redirect you to AMO, which should redirect you back to FxA asking you to set up a 2nd factor.
- You should be redirected back to AMO once the 2nd factor is set up
- You shouldn't need to enter your FxA credentials twice.
- Once a 2nd factor is set up, FxA should ask for it every time you log back to AMO.

**Comment by AlexandraMoga** [comment]
@diox I've tested the first part - without the waffle flag enabled - and the login and register process continues to work as before.
For the second scenario - with the waffle flag active - I am not getting at the expected results mentioned above, i.e I'm not being redirected to set up 2FA for my developer account once I enter the password. I'm being logged in normally. 


Here is the [waffle flag](https://addons-internal.dev.mozaws.net/en-US/admin/models/waffle/flag/20/change/) I've added in admin. Maybe I messed up the set up somehow.

**Comment by diox** [comment]
The flag looks ok, and I can see it being enabled for your user. I was expecting breakage, but not this... I'll test further, as I said above it doesn't block verification of this issue as we are not likely to enable the flag for everyone in prod soon.

**Comment by diox** [comment]
@AlexandraMoga heads up, the flag should be fixed on dev soon, I just merged a patch for it.

**Comment by AlexandraMoga** [comment]
@diox I'm not able to log in at all with the user added in the [flag](https://addons-internal.dev.mozaws.net/en-US/admin/models/waffle/flag/20/change/)

**Comment by diox** [comment]
Yes, same here, FxA calls our authentication callback with `error=unauthorized_client`. I mentioned this in the issue on their end: https://github.com/mozilla/fxa/issues/2024#issuecomment-662935099

**Comment by AlexandraMoga** [comment]
@diox here is a list of scenarios that I've been trying out to verify the fix for https://github.com/mozilla/fxa/issues/6342

**Working as expected**
- a recent developer account added to the enforced 2FA flag will be asked to set up 2FA at login; every subsequent login will require a 2FA code
- a developer account with enforced 2FA changes their primary email address; 2FA code is still required when they log in to AMO with the updated primary email 
- a reviewer/developer account with prior 2FA added to the enforced 2FA user list will continue to be asked to enter a 2FA code at log in, i.e. they are not asked to set up 2FA again
- theme artists and regular users are not asked to set up 2FA 

**Old accounts**
- the tests done with my old reviewer and developer accounts still reproduced the "Authorize this sign in" page (see screenshot). This happened when I've tried to log in after I've added those accounts to the enforced 2FA flag.
![image](https://user-images.githubusercontent.com/31961530/93364207-f2097680-f850-11ea-8f31-540728756a11.png)


I've also filed #15505 as a follow up.

**Comment by diox** [comment]
Were you still stuck at the `Authorize this sign-in` screen, or were you able to authorize it by pasting the correct code sent to your email ?

**Comment by AlexandraMoga** [comment]
> Were you still stuck at the `Authorize this sign-in` screen (as was happening to me in [mozilla/fxa#6342](https://github.com/mozilla/fxa/issues/6342)), or were you able to authorize it by pasting the correct code sent to your email ?

I wasn't stuck at this step. I used the code received on my email and then I was also prompted to set up 2FA, so at least that works. Once 2FA was active for these accounts, the "Authorize this sign-in" screen no longer appears at log in. 
I raised this issue because I was under the impression that the "Authorize this sign-in" screen should not appear at all.
